### PR TITLE
Create the emitter and listener servers

### PR DESCRIPTION
**Pull request by mcolmant**
* cmx-emitter: NodeJS server for emitting IndoorLocation objects retrieved from redis notifications into sockets
* cmx-listener: NodeJS server for listening CMX notifications, process them to IndoorLocation objects  and save them to redis

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,66 @@
 # cmx-socket-indoor-location-emitter
-Node.js server to provide IndoorLocation from Cisco CMX. Works with socketIndoorLocationProviders libraries on iOS, Android and JS.
+Node.js server to provide IndoorLocation from Cisco CMX.
+Works with socketIndoorLocationProviders libraries on iOS, Android and JS.
```
Comment (mathieugerard) [inline README.md]: socketIndoorLocationProvider

File: README.md [hunk 2]
```diff
@@ -1,2 +1,66 @@
...
+*   Clone the repository
+*   Go to this folder, and install the modules
+    ```
+    npm i
+    ```
+
+## cmx-listener
+
+NodeJS server to react on CMX notifications.
+Each notification will be received and processed into an IndorLocation object.
```
Comment (mathieugerard) [inline README.md]: IndoorLocation

File: README.md [hunk 3]
```diff
@@ -1,2 +1,66 @@
...
+The notification will be saved into a MongoDB collection for later used if enabled, the IndoorLocation object into a Redis database instead.
```
Comment (mathieugerard) [inline README.md]: The computed indoorLocation will be saved in a redis database with the IP address as userId to be used by the emitter.

The notification can also be stored into a MongoDB collection for later use if enabled.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,66 @@
...
+        *   SESSIONS_REDIS_AUTH: redis password (if set)
+        *   REDIS_CMX_NOTIF_TTL: redis key TTL
+        *   MONGODB_ENABLED: enable notifications logging into MongoDB
+        *   MONGODB_URL: MongoDB URL
+        *   MONGODB_COLLECTION: MongoDB collection to use
+*   Start the server
+    ```
+    npm run start-listener
+    ```
+
```
Comment (mathieugerard) [inline README.md]: ## CMX Notification configuration

In your CMX dashboard, go to Manage/Notifications. Create a notification with the following parameters:

- type = "location update" or "movement". If you select movement, set a distance like 3 or 5 feet.
- receiver
  * protocol = http or https depending on your deployment
  * host = the host where the cmxlistener is deployed
  * port = 80 or 443 to match http or https
  * url = "?k={KEY}" where {KEY} is your KEY environment variable

You can use https://cmxlocationsandbox.cisco.com, with learning/learning as credentials, to test your service.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,66 @@
...
+## cmx-emitter
+
+NodeJS server to react on socket indoor location providers.
+Each socket connexion will lead to a redis subscription that will helps to only get the IndoorLocation objects when the location of a user has been changed.
```
Comment (mathieugerard) [inline README.md]: connection
help

File: cmx-emitter/config/config.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Load app configurations
+ */
+module.exports = require('./env/all');
+
+/**
+ * Get files by glob patterns
+ */
+module.exports.getGlobbedFiles = function (globPatterns, removeRoot) {
```
Comment (mathieugerard) [inline cmx-emitter/config/config.js]: Used ?

File: cmx-emitter/config/express.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+module.exports = function () {
+
+    var app = require('express')();
+
+    var server = require('http').Server(app);
+    var io = require('socket.io')(server);
+
+    // Express configuration
+    app.use(bodyParser.json());
```
Comment (mathieugerard) [inline cmx-emitter/config/express.js]: not needed

File: cmx-listener/config/config.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Load app configurations
+ */
+module.exports = require('./env/all');
+
+/**
+ * Get files by glob patterns
+ */
+module.exports.getGlobbedFiles = function (globPatterns, removeRoot) {
```
Comment (mathieugerard) [inline cmx-listener/config/config.js]: used?

File: cmx-listener/config/env/all.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    redis: {
+        port: process.env.SESSIONS_REDIS_PORT || '6379',
+        host: process.env.SESSIONS_REDIS_HOST || 'localhost',
+        password: process.env.SESSIONS_REDIS_AUTH,
+        cmxNotificationTTL: process.env.REDIS_CMX_NOTIF_TTL || 3600,
+    },
+    mongodb: {
+        url: process.env.MONGODB_URL || 'mongodb://localhost:27017/mapwize-cmx-logger',
+        collection: process.env.MONGODB_COLLECTION || 'logs',
+        enabled: process.env.MONGODB_ENABLED || true
```
Comment (mathieugerard) [inline cmx-listener/config/env/all.js]: Should be disabled by default?

File: cmx-listener/fake-cmx-notifification.json [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+            "floorId" : 723413320329068700,
+            "floorRefId" : null,
+            "entity" : "WIRELESS_CLIENTS",
+            "deviceId" : "00:00:2a:01:00:0a",
+            "lastSeen" : "2017-10-18T13:12:19.838+00:00",
+            "moveDistanceInFt" : 56.645954,
+            "timestamp" : 1508332339838.0
+        }
+    ],
+    "publicIp" : "172.22.0.1",
```
Comment (mathieugerard) [inline cmx-listener/fake-cmx-notifification.json]: Remove publicIp and timestamp.

Move this to a test directory ?

File: cmx-listener/utils/mapwize.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
+'use strict';
+
+var config = require('../config/config');
+var rawLayers = require('../raw-layers.json');
```
Comment (mathieugerard) [inline cmx-listener/utils/mapwize.js]: We need to check how to best set floor projection parameters

File: README.md (hunk 4 above)
Comment (mcolmant) [reply README.md]: Done.

**Pull request closed without merging**